10/25 Year Loan toggle added

mistergone opened the pull request:
This PR adds the 10/25 year loan term toggle to the appropriate metrics sections. I also went down the rabbit hole a bit and refactored `metrics-view`. **Note**: The styles are _mostly_ correct, but there are slight differences between the final design and the current look of the app.
## Additions
- 'How much will I pay per month for all of my student loans?' section
- Radio buttons in two sidebars of Step 2
- Handlers which update various values based on a 10 or 25-year loan term
## Changes
- `metric-view` refactored
  - Added an object property, `metrics`, which stores important numbers for later use
  - Refactored methods to use fewer parameters
  - Removed and/or changed the way several methods work
## Testing
- The app passes both existing and new functional and unit tests. (See the README for how to test.)
## Review
- @marteki and @higs4281 
- @niqjohnson - If you have some time, I'd love for you to check this out. I rewrote a lot of `metrics-view`, and if you can check whether I missed something, that'd be great!
## Screenshots

<img width="689" alt="screen shot 2016-06-09 at 11 22 11 am" src="https://cloud.githubusercontent.com/assets/1490703/15935192/76f74dcc-2e34-11e6-80d4-dd3afdb90ff9.png">
## Checklist
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows the standards laid out in the [front end playbook](https://github.com/cfpb/front-end)
- [x] Passes all existing automated tests
- [x] New functions include new tests
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged
- [x] Visually tested in supported browsers and devices
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
## Animated gif

**"Really, Chuck. You think now is a good time to refactor?"**

![spock-eye-raise](https://cloud.githubusercontent.com/assets/1490703/15935252/b06e1432-2e34-11e6-8c83-6d70f30f7205.gif)

mistergone committed:
This commit adds the necessary HTML markup for 10/25 year repayment toggles to
be added to the front-end. It also cleans up some things in metric-view that
were a bit long-winded for my tastes. This is in preparation to rewrite
portions of metric-view to more easily handle the 10/25 year repayment
toggles.

mistergone committed:
As of this commit, the 10/25 year loan term toggles are functional.
However, notifications for debt burden are not. I intend to refactor
the notifications code in metric-view to eliminate the need for
functions that take several parameters, and to take data about limits
and metrics out of the HTML and place them in Objects inside metric
view, where they'll be easier to find if they need to be updated.

mistergone committed:
Refactor of metric-view - Many of the functions in this file had up to 5
parameters, and when I started the work on the 10/25 year toggle, I thought I
might be able to simplify them. In many cases, all the same data (the school
and national metrics) were being used by many different methods, so I put that
data into a property of metric-view called 'metrics'. In many cases, this
allowed me to change function calls which wrapped over two lines into more
readable, single-line calls. I also updated the metric-view tests to match the
changes to these methods.

mistergone committed:
CHANGELOG updated

mistergone committed:
Added missing bit of HTML content

marteki commented on src/disclosures/js/models/school-model.js:
```diff
@@ -18,7 +18,8 @@ var schoolModel = {
     values.jobRate = values.jobRate || '';
     values.programLength /= 12;
-    values.medianSalary = values.salary || values.medianAnnualPay;
+    values.medianSalary = values.salary || values.medianAnnualPay ||
```
All the other lines in this sections seem to be first looking for program level data, then school level data. This one is comparing three values.

Can you talk a little more about what's now going on for the medianSalary value, or point me at the documentation about it? The variable names are so similar, and it's hard to know where each datapoint is coming from or represents. 😕

marteki commented on src/disclosures/js/views/metric-view.js:
```diff
@@ -5,94 +5,113 @@ var getSchool = require( '../dispatchers/get-school-values' );
 var metricView = {
+  metrics: {
+    debtBurden: {
+      school: NaN,
```
I have a few guesses, but rather than assume, 😄  I'll ask: why are we explicitly initializing these as NaNs?

mistergone replied:
I think it's best to assume there's no data available, which means no number (rather than showing an inaccurate default number).

marteki commented on test/js-unit/metric-view-spec.js:
```diff
@@ -4,151 +4,90 @@ var metricView = require( '../../src/disclosures/js/views/metric-view' );
+  it( 'gives the proper notification classes when min and max ' +
+       'values are not numbers', function() {
+    metricView.metrics.test.national = 0.5;
+    metricView.metrics.test.school = 0.5;
```
Why did the school value increase here? Can we change this back to 0.3, to match the previous version of the test (refactoring it rather than testing something new)?

mistergone replied:
I don't understand your concern. We should talk about it. :)

marteki commented on src/disclosures/js/views/metric-view.js:
```diff
@@ -107,86 +126,94 @@ var metricView = {
-  getNotificationClasses: function( schoolValue, nationalValue, sameMin,
-    sameMax, betterDirection ) {
-    var notificationClasses = '';
+  getNotifications: function( metricKey ) {
```
> gulp lint
>   190:21  warning  Function 'getNotifications' has a complexity of 11

😲 

The code refactor is _increasing_ the cyclomatic complexity of this section, and into the double digits. As much as I enjoy Spinal Tap, I'm pretty uncomfortable with the idea of going to 11. Could we use a different approach in the code to keep from doing that?

mistergone replied:
The complexity number is basically the number of `if` statements in the function. This function isn't a bunch of scary nested `if` statements, but rather a series of `if - else` statements checking the state of the school and national numbers. We could certainly refactor the code, but I'm not convinced it would make it easier to read or more efficient. I'm open to ideas, though!

higs4281 commented:
Code works for me, but it leaves me with a question:  
When a student chooses the 25-year option at "Calculate your monthly payment," shouldn't that choice propagate to the "Your estimated debt burden" section, and vice versa? Otherwise, you have a possibly inadvertent mismatch.

Other than that, 👍

mistergone commented:
@higs4281 - The original specification was that each section was independent of the other. If you think it's an issue, we can see how everyone (@kurzn @marteki and perhaps @niqjohnson and @caheberer) feels about the final interactions and revise them if necessary.

higs4281 commented:
👍  agreed -- question is outside scope of this PR.

mistergone committed:
To remove the complexity warning for getNotifications(), I refactored this
file a bit. Instead of getNotifications() doing the entire lift for
determining the 'standing' (higher, lower, about the same) of a school per
metric, I added a function, checkMetrics(), which does a simple set of checks
of the school values against the high and low to determine the 'standing'
property for that metric. These standings are then used by getNotifications()
to assign classes. getNotifications() also checks if any data is missing, and
returns the proper classes based on all these checks.

With these changes, I was able to completely eliminate the complexity warning
we were getting for getNotifications().

mistergone commented:
@marteki - I addressed the complexity issue with some refactors which may not make the whole thing easier to read (it's debatable!) but definitely makes it more atomic, with less individual function complexity. Here's the commit message for details:

> To remove the complexity warning for getNotifications(), I refactored this file a bit. Instead of getNotifications() doing the entire lift for determining the 'standing' (higher, lower, about the same) of a school per metric, I added a function, checkMetrics(), which does a simple set of checks of the school values against the high and low to determine the 'standing' property for that metric. These standings are then used by getNotifications() to assign classes. getNotifications() also checks if any data is missing, and returns the proper classes based on all these checks.
> 
> With these changes, I was able to completely eliminate the complexity warning we were getting for getNotifications().

This commit still passes all existing tests. I added five new tests for the new function, `checkMetrics()`, in `/tests/js-unit/metric-view-spec.js`

marteki commented:
![94dnw](https://cloud.githubusercontent.com/assets/1183435/16021259/a69838ea-317f-11e6-9d67-56058a8dbe04.gif)

Thank you for looking into the complexity!

:+1: from me for merge.

caheberer commented:
I'm not sure if now is the time to say that the radio button selection state doesn't match our standards...but it should look like this:

![screen shot 2016-06-13 at 4 04 03 pm](https://cloud.githubusercontent.com/assets/4739887/16021419/99dc9a82-3180-11e6-9616-6fd1e6171a62.png)

But in regards to why I was @'ed. From what I recall, we agreed the toggle wouldn't change numbers in other sections at this time.

mistergone commented:
@marteki  - I see your Pinkie Pie and raise you Cheese Sandwich

<img src="http://img0.joyreactor.cc/pics/post/my-little-pony-mlp-%D0%BF%D0%B5%D1%81%D0%BE%D1%87%D0%BD%D0%B8%D1%86%D0%B0-%D1%84%D1%8D%D0%BD%D0%B4%D0%BE%D0%BC%D1%8B-mlp-1056628.gif">

niqjohnson commented:
Looks like a solid refactor to me. Nice!

Only thing I noticed is that the JSdoc comments for some of the refactored methods (`setGraphValues`, `setGraphPositions`) still have some old `@param` notes in them.

mistergone commented:
@caheberer - I noted it in the Pull Request description, but you probably missed it. This PR is mostly for the front-end functional changes, so the styles are not 100% to spec yet.

caheberer commented:
@mistergone 
![giphy](https://cloud.githubusercontent.com/assets/4739887/16021542/487737dc-3181-11e6-9aa6-cd512833232c.gif)

mistergone commented:
@niqjohnson - Good catch! In the refactor, I tried to update JSDocs as I went, but I obviously missed some!

mistergone committed:
Removed out-of-date JSDOC comments
